controllers/gitrepository_controller.go

```diff
@@ -703,13 +712,25 @@ func (r *GitRepositoryReconciler) garbageCollect(ctx context.Context, obj *sourc
 		return nil
 	}
 	if obj.GetArtifact() != nil {
-		if deleted, err := r.Storage.RemoveAllButCurrent(*obj.GetArtifact()); err != nil {
-			return &serror.Event{
-				Err: fmt.Errorf("garbage collection of old artifacts failed: %w", err),
+		delFilesChan := make(chan []string)
+		errChan := make(chan error)
+		go r.Storage.GarbageCollect(*obj.GetArtifact(), delFilesChan, errChan)
+		for {
+			select {
+			case <-ctx.Done():
+				return fmt.Errorf("context deadline exceeded")
```

Comment:
It may be worth adding some contextual information:
```suggestion
				return fmt.Errorf("garbage collection: context deadline exceeded")
```